actors/builtin/market/market_actor.go

```diff
@@ -525,7 +532,7 @@ func (a Actor) CronTick(rt Runtime, params *adt.EmptyValue) *adt.EmptyValue {
 					amountSlashed = big.Add(amountSlashed, slashAmount)
 				}
 
-				if nextEpoch != epochUndefined {
+				if nextEpoch != epochUndefined && !removeDeal {
```

Comment:
My main concern with this structure is not that #766 could actually happen, it's that the branching logic does not correspond to the range of returned values from `updatePendingDealState`. Here's what I would like to see:

```go
slashAmount, nextEpoch, removeDeal := msm.updatePendingDealState(...)
// Should always hold
Assert(slashAmount >= 0)
if removeDeal {
  // If we're removing the deal, there is no "next epoch:"
  Assert(nextEpoch == epochUndefined)

  // Handle everSlashed case:
  if slashAmount != 0 {
    amountSlashed = big.Add(amountSlashed, slashAmount)
  }

  // Delete deal from Proposals and DealStates:
  if err := deleteDealProposalAndState(...); err != nil {
    builtin.RequireNoErr(rt, err, ...)
  }
} else {
  Assert(nextEpoch > rt.CurrEpoch() && amountSlashed == 0)

  // Update deal's LastUpdatedEpoch in DealStates
  state.LastUpdatedEpoch = rt.CurrEpoch()
  if err := msm.dealStates.Set(dealID, state); err != nil {
    rt.Abortf(...)
  }

  // Add dealID to updatesNeeded for re-insertion in DealOpsByEpoch
  updatesNeeded[nextEpoch] = append(updatesNeeded[nextEpoch], dealID)
}
```

I'm not crazy about adding `Asserts`, but my understanding of the market and `updatePendingDealStates` is that these should all be fine.

Reply:
I see what you mean now. Thanks.